### ISSUE TITLE
Typed NamedTuple with member function generating another instance of class fails at runtime with NameError

### ISSUE THREAD

**Issue by lwolfsonkin** [python/typing#485]
I simplified the code just for proof of concept:

```python
class Hypothesis(NamedTuple):
    logprob: float
    def attach(self) -> Hypothesis:
        return Hypothesis(logprob=4)
```

While `mypy` doesn't complain at this, the python interpreter crashes with a `NameError` saying:

```
Traceback (most recent call last):
  File "./decode.py", line 31, in <module>
    class Hypothesis(NamedTuple):
  File "./decode.py", line 34, in Hypothesis
    def attach(self) -> Hypothesis:
NameError: name 'Hypothesis' is not defined
```

Unfortunately, I dropped the return type annotation as a result (which mypy didn't complain about) in order to make it runnable.

**Comment by ethanhs** [comment]
This is because the name `Hyposthesis` is not defined inside its own class definition body. Mypy accepts this due to supporting this in stubs being needed. You can change the return annotation to be `-> 'Hypothesis'` and it will both work at runtime and be understood by mypy.

**Comment by lwolfsonkin** [comment]
Thanks for the tip! Interesting. So, the name `Hypothesis` is not defined in the class body, but is defined in methods within that class body?

**Comment by ethanhs** [comment]
Methods are part of the class body. So this also will not work

```python
class Example:
    def test(a: Example):
        pass
```

nor this

```python
class Example:
    def test(a) -> Example:
        pass
```

**Comment by lwolfsonkin** [comment]
huh. Cool. Definitely good to know. Thanks a lot @ethanhs!

**Comment by gvanrossum** [comment]
I do think there's a mypy bug here -- outside stubs, this should be rejected.

**Comment by JukkaL** [comment]
> I do think there's a mypy bug here -- outside stubs, this should be rejected.

Yes, and we already have a mypy issue about this (https://github.com/python/mypy/issues/948).

**Comment by ilevkivskyi** [comment]
@gvanrossum 
> I do think there's a mypy bug here -- outside stubs, this should be rejected.

...or PEP 563 can be accepted :-)
